lib/build/Dockerfile

```diff
@@ -1,19 +1,21 @@
-FROM ubuntu:18.04
-RUN apt-get update && \
-    apt-get install -y apt-transport-https git gnupg software-properties-common wget && \
-    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - && \
-    add-apt-repository "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-12 main" && \
-    apt-get update && \
-    apt-get install -y clang-12 make && \
-    apt-get remove --purge -y software-properties-common wget && \
-    apt-get autoremove --purge -y && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/*
-
+FROM alpine:3.14.2 as base
 ENV HOME /home/me
+RUN apk add --no-cache binutils
+COPY --from=symbolica/clang-builder /tmp/clang/install/bin/clang-12 /usr/local/bin/
+COPY --from=symbolica/clang-builder /tmp/clang/install/bin/llvm-link /usr/local/bin/
```

Comment:
So the approach I'm taking here is to copy just the relevant binaries for the `clang-builder` container. This means we need to make sure this one is built first so that this and translate image can consume it. My thinking here was that by combining all the clang compilation into one Docker file it would reduce duplication, both in code and in time. However, I think the time saving is minimal because it's compilation of `llvm-extract` and `clang` that take up 99% off the time here (clang being the worse offender). So I think actually, if we just have a script called `clang-cmake.sh` or something, which contains all the common bits up to the point we call `ninja` then we can inline the other bits into the `build` and `translate` images and reuse that script there as those two images don't need any of the same compiled llvm/clang binaries anyway. 

Also, once the translate stuff is sorted then we can just inline this all into the `build` image and do away with the compilation of `llvm-extract` and `opt` altogether.

Reply:
In fact, there is some time saving by combining them like this because a good chunk of the cpp.o files are used by more than one of these tools and so by the time it gets around to build `clang` it's only got 994 objects left to build, vs (I think) 2300 if it were to build clang in isolation. Still, given that these base layers shouldn't change too frequently and that translate will go soon, I'm still probably veering towards creating the shared script approach rather than the shared base image approach.